modules/tensor_mechanics/src/kernels/TotalLagrangianStressDivergence.C

```diff
@@ -0,0 +1,209 @@
+#include "TotalLagrangianStressDivergence.h"
+
+#include "HomogenizationConstraintIntegral.h"
+
+registerMooseObject("TensorMechanicsApp", TotalLagrangianStressDivergence);
+
+InputParameters
+TotalLagrangianStressDivergence::validParams()
+{
+  InputParameters params = Kernel::validParams();
+
+  params.addRequiredParam<unsigned int>("component", "Which direction this kernel acts in");
+  params.addRequiredCoupledVar("displacements", "The displacement components");
+  params.addParam<bool>("large_kinematics", false, "Use large displacement kinematics");
+
+  // This kernel requires use_displaced_mesh to be off
+  params.suppressParameter<bool>("use_displaced_mesh");
+
+  params.addCoupledVar("macro_gradient", "Optional scalar field with the macro gradient");
+
+  params.addParam<MultiMooseEnum>("constraint_types",
+                                  HomogenizationConstants::mooseConstraintType,
+                                  "Type of each constraint: strain or stress");
+
+  return params;
+}
+
+TotalLagrangianStressDivergence::TotalLagrangianStressDivergence(const InputParameters & parameters)
+  : DerivativeMaterialInterface<Kernel>(parameters),
+    _ld(getParam<bool>("large_kinematics")),
+    _component(getParam<unsigned int>("component")),
+    _ndisp(coupledComponents("displacements")),
+    _disp_nums(_ndisp),
+    _disp_vars(_ndisp),
+    _pk1(getMaterialPropertyByName<RankTwoTensor>("pk1_stress")),
+    _dpk1(getMaterialPropertyByName<RankFourTensor>("pk1_jacobian")),
+    _macro_gradient_num(isCoupledScalar("macro_gradient", 0) ? coupledScalar("macro_gradient") : 0),
+    _macro_gradient(isCoupledScalar("macro_gradient", 0) ? getScalarVar("macro_gradient", 0)
+                                                         : nullptr),
```

Comment:
```suggestion
    _macro_gradient(isParamValid("macro_gradient") ? getScalarVar("macro_gradient", 0)
                                                         : nullptr),
```

Reply:
Actually why don't you directly get the variable order here, because that's all you need for the jacobian calculations, right?

Reply:
Yes, this is a much better way to do it!